:memo: Updated Getting Started / Concepts (#4)
writechoiceorg/terrateam: docs/src/content/docs/getting-started/concepts.mdx

> ## Dirspace

Reviewer: remove the `## Key Concepts and Terminology` title from the previous section.

> At the core of Terrateam is the concept of a Dirspace. It represents a unique blend of a repository, directory, and workspace. It is referred to as a tuple `(repository, directory, workspace)` or simply `(directory, workspace)` when the repository is implied.

Reviewer: What do you mean by "repository is implied."? It's not clear

> At the core of Terrateam is the concept of a Dirspace. It represents a unique blend of a repository, directory, and workspace. It is referred to as a tuple `(repository, directory, workspace)` or simply `(directory, workspace)` when the repository is implied.

Reviewer: ```suggestion
In Terrateam, a Dirspace is a fundamental concept that defines the scope of Terraform operations. It uniquely combines a `repository`, a `directory` within that repository, and a Terraform `workspace`, represented as a tuple `(repository, directory, workspace)`. When the repository context is clear, it can be simplified to `(directory, workspace)`.
```

> Dirspaces are automatically created when Terrateam detects Terraform configurations in your repository. Each directory containing Terraform files forms a Dirspace with either its explicitly configured workspace or the **default workspace** (created automatically when no specific workspace is defined).

Reviewer: ```suggestion
Dirspaces are automatically created when Terrateam detects Terraform configurations in your repository. Each directory containing Terraform files creates a Dirspace with either its explicitly configured workspace or the **default workspace** (created automatically when no specific workspace is defined).
```

#### **Example:**

Imagine you have a repository called `infra-repo` managing multiple environments:

> | Repository | Directory | Workspace |

Reviewer: Let's use a diagram:

![image](https://github.com/user-attachments/assets/beae45b2-16b9-42ce-bb43-05831efa2c9f)

Reviewer: Here's the code:
```

graph TD;
    subgraph Repository["Repository"]
        infra["infra-repo"]
    end

    subgraph Directory["Directory"]
        dev["dev"]
        staging["staging"]
        prod["prod"]
    end

    subgraph Workflow["Workflow"]
        default_ws["default"]
    end

    infra --> dev
    infra --> staging
    infra --> prod
    
    dev --> default_ws
    staging --> default_ws
    prod --> default_ws

```

| ---------- | --------- | --------- |
| infra-repo | dev       | default   |
| infra-repo | staging   | default   |
| infra-repo | prod      | default   |

A **workspace** enables you to manage multiple states for the same directory, isolating environments like `dev`, `staging`, and `prod` within the same repository.

```
Repository: infra-repo
  |
  |-- dev/
  |    |-- Workspace: default
  |
  |-- staging/
  |    |-- Workspace: default
  |
  |-- prod/
       |-- Workspace: default
```
:::note
Terrateam executes operations like plan and apply within the context of a Dirspace.
:::
## Changes and File Patterns

> When a pull request is opened or updated, Terrateam detects changes based on patterns defined in your `.terrateam/config.yml` file under the `when_modified` configuration. The default configuration looks like:

Reviewer: ```suggestion
When a pull request is opened or updated, Terrateam detects changes based on patterns defined in your `.terrateam/config.yml` file under the [`when_modified`](/configuration-reference/when-modified) configuration. The default configuration looks like:
```

Reviewer: Is the Terrateam/config file automatically generated when the terrateam app is installed? Until now, I didn't know about this file. We need to clarify this.

```yaml
when_modified:
  file_patterns:
    - '**/*.tf'
    - '**/*.tfvars'
```

This configuration instructs Terrateam to monitor changes to any files with `.tf` or `.tfvars` extensions. Adjust these patterns in your configuration file to align with your project's structure.

Terrateam maps detected changes to their appropriate [Dirspaces](/getting-started/concepts#dirspace). These mapped Dirspaces are referred to as Changes. When executing operations, Terrateam only considers Dirspaces with modified files in the pull request.

## Auto-Plan and Auto-Apply

> Terrateam provides automated Terraform execution workflows to simplify infrastructure changes:

Reviewer: Please, inform in which file the following configurations are available.

- **Auto-Plan**: Runs `terraform plan` automatically when a pull request is created or updated. Enable with:
  ```yaml
  auto_plan: true  # Set to false to disable
  ```

> - **Auto-Apply**: (Disabled by default) Runs `terraform apply` post-merge. Enable with:

Reviewer: ```suggestion
- **Auto-Apply**: Runs `terraform apply` post-merge. It is disabled by default. Enable with:
```

  ```yaml
  auto_apply: true
  ```

> These automated operations streamline workflows while maintaining control through configurable thresholds.

Reviewer: Is there any other page that address the Plan or Apply? You can add a note link to these complementary materials.

> ## Tags and Tag Sets

Reviewer: It's not clear where or why we need to use Tags. Can you improve the explanation? If is there another page covering this, please, add a link to it.

Reviewer: Who are responsible for adding tags? The user? Or are tags added automatically?

Terrateam uses Tags to organize Terraform resources through flexible labeling:

* **Tag Set**: Unordered, unique labels assigned to resources  
  Example: `env:prod, component:network`
* **Tag Query**: Filter to match specific Tag Sets  
  Example: `env:prod` matches all production resources

> **Example workflow:**

Reviewer: It's not a title, don't use bold. Change it to a normal sentense, like, The following steps describe the Tags workflow (need to check).

1. Tag EC2 instances with `resource-type:compute`
2. Create Auto-Scaling Groups with `resource-type:compute, environment:production`
3. Query with `resource-type:compute` to manage all compute resources

Tags enable precise resource management across complex infrastructures.
## Locks and Unlocks

Terrateam's locking system prevents concurrent modifications by ensuring only one change can be applied to a resource at a time. This prevents:

- State file conflicts  
- Resource modification races  
- Concurrent apply operations  

> **Key behaviors**:

Reviewer: It should not be bold, as described before. Add as a normal sentence.

> - Enabled by default (no config needed)

Reviewer: Add endpoint to sentences.

- Automatically released after apply + merge
- Force unlock via comment: `terrateam unlock`

> ```yaml

Reviewer: Before presenting the code snippet, add a sentence describing it. By the way, in the code block, the property is called lock_policy, therefore, we need to describe it as a Lock Policy. It probably is a better tytle compare to Locks and Unlocks.

# .terrateam/config.yml (default shown)
workflows:
  lock_policy: strict  # strict|apply|merge|none
```

For advanced configurations see our [Locks & Concurrency guide](/advanced-workflows/locks-and-concurrency).

> :::note

Reviewer: Move this note to the top of the page, just after the introduction.

> At the heart of Terrateam is the concept of a Dirspace. A Dirspace represents a unique combination of a repository, a directory, and a workspace. It is denoted as a tuple `(repository, directory, workspace)` or simply `(directory, workspace)` when the repository is implied.

Reviewer 2: Dirspaces are introduced abstractly; provide a real-world example or use case before delving into the technical details.

Reviewer 2: > Dirspaces are introduced abstractly; provide a real-world example or use case before delving into the technical details.

```suggestion
At the core of Terrateam is the concept of a Dirspace. It represents a unique blend of a repository, directory, and workspace. It is referred to as a tuple `(repository, directory, workspace)` or simply `(directory, workspace)` when the repository is implied.
```

Dirspaces are automatically created when Terrateam detects Terraform configurations in your repository. Each directory containing Terraform files forms a Dirspace with either its explicitly configured workspace or the **default workspace** (created automatically when no specific workspace is defined).

#### **Example:**
Consider a repository named `repo` with two directories, `dir1` and `dir2`, and no explicitly configured workspaces. This repository would have two Dirspaces:

| Repository  | Directory | Workspace |
|------------|----------|-----------|
| repo       | dir1     | default   |
| repo       | dir2     | default   |

> A **workspace** allows you to manage multiple states for the same directory. This helps isolate environments such as `dev`, `staging`, and `prod` within the same repository, making infrastructure management more organized.

Reviewer 2: ```suggestion
A **workspace** enables you to manage multiple states for the same directory. This helps to isolate environments such as `dev`, `staging`, and `prod` within the same repository, making infrastructure management more organized.
```

  |-- Directory 2
        |-- Workspace: default
```
:::note

> Terrateam executes operations like plan and apply within the context of a Dirspace.

Reviewer 2: "Terrateam executes" appears three times in the text. You could mention it once, then reference it in the later sections.

:::
## Changes and File Patterns

When a pull request is opened or updated, Terrateam detects changes based on patterns defined in your `.terrateam/config.yml` file under the `when_modified` configuration. The default configuration looks like:

```yaml
when_modified:
  file_patterns:
    - '**/*.tf'
    - '**/*.tfvars'
```

> This configuration tells Terrateam to monitor changes to any files with `.tf` or `.tfvars` extensions. Modify these patterns in your configuration file to match your project's structure.

Reviewer 2: ```suggestion
This configuration instructs Terrateam to monitor changes to any files with `.tf` or `.tfvars` extensions. Adjust these patterns in your configuration file to align with your project's structure.
```

Terrateam maps detected changes to their appropriate [Dirspaces](/getting-started/concepts#dirspace). These mapped Dirspaces are referred to as Changes. When executing operations, Terrateam only considers Dirspaces with modified files in the pull request.

> ## Auto-Plan and Auto-Apply

Reviewer 2: Auto-Plan and Auto-Apply could benefit from a brief summary. Please add a one-line description before the configuration example for clarity.

Terrateam offers two configurable automated operations:New chat
  auto_plan: true  # Set to false to disable
  ```
- **Auto-Apply**: (Disabled by default) Runs `terraform apply` post-merge. Enable with:
  ```yaml
  auto_apply: true
  ```

These automated operations streamline workflows while maintaining control through configurable thresholds.

## Locks and Unlocks

Terrateam's locking system prevents concurrent modifications by ensuring only one change can be applied to a resource at a time. This prevents:

- State file conflicts  
- Resource modification races  
- Concurrent apply operations  

**Key behaviors**:
- Enabled by default (no config needed)
- Automatically released after apply + merge
- Force unlock via comment: `terrateam unlock`

```yaml
# .terrateam/config.yml (default shown)
workflows:
  lock_policy: strict  # strict|apply|merge|none
```

For advanced configurations see our [Locks & Concurrency guide](/advanced-workflows/locks-and-concurrency).

:::note
 Default settings work for most teams. Customize only if you need specific lock timing or policies.
:::

> ## Tags and Tag Sets

Reviewer 2: Tags and tag queries are introduced a little late. Moving them higher in the document, preferably before locks, would be a good idea.

Terrateam uses Tags to organize Terraform resources through flexible labeling:

* **Tag Set**: Unordered, unique labels assigned to resources  
  Example: `env:prod, component:network`
* **Tag Query**: Filter to match specific Tag Sets  
  Example: `env:prod` matches all production resources

**Example workflow:**
1. Tag EC2 instances with `resource-type:compute`
2. Create Auto-Scaling Groups with `resource-type:compute, environment:production`
3. Query with `resource-type:compute` to manage all compute resources

> Tags enable precise resource management across complex infrastructures.

Reviewer 2: ```suggestion
Tags facilitate effective resource management in complex environments infrastructures.
```
